### ISSUE TITLE
[Bug] Privacy issue: dataset and metadata of dataset

### ISSUE THREAD

**Issue by XinYao1994** [TL-System/plato#97]
**Describe the bug**
https://github.com/TL-System/plato/blob/f9a8f7758536f47de14130c19b99a0c128440a40/plato/algorithms/tensorflow/fedavg.py#L22-L23

**Expected behavior**
In Federated Learning, the server should not directly access the data locally.
This code assumes the server can access all the data when it creates some objects. Maybe we can distinguish the dataset from its metadata (number of classes, labels, etc).

**Additional context**
This is not a runtime bug.

**Comment by baochunli** [comment]
This is a known issue that at the time of writing this code, I am not sure how to solve this appropriately. TensorFlow (Keras) models need to be built before being used, but to build the model, one needs the input shape of the dataset. The `DataSource` class then needs to include an input_shape() function to return the input shape of the dataset. The `datasource` instance, however, doesn't need to load any data when initializing itself. Conceptually, one can detach all the data loading code from `init()`, and put them in a separate function.

**Comment by XinYao1994** [comment]
Similar things happen when using yolov5 in Pytorch. The input shape of a model should be better defined in a model configuration, instead of being dynamicly adjusted based on the given dataset.

**Comment by baochunli** [comment]
It makes sense to say that the model should somehow provide the input shape, however since we are using vanilla models from PyTorch (`nn.Model`), MindSpore (`nn.cell`), and TensorFlow/Keras, I am not sure how this can be obtained. One possibility, which is a bit less elegant, is to explicitly add the data shape into the configuration file.

The Yolov5 + coco128 work is done jointly with Qintao. Perhaps he can take a look at this issue -- assigned.

**Comment by TrackerLB** [comment]
Perhaps  define a configuration to solve this issue.
